### ISSUE TITLE
Issues with Postman's Global/System Proxy

### ISSUE THREAD

**Issue by xk0der** [a85/POSTMan-Chrome-Extension#7676]
### **Describe the bug**
**Issue 1**
Say if a Windows system is not domain joined but has to rely on a Proxy server that expects the username to be in the format `domain\username` - The Global proxy is trying to resolve the `domain` component and failing with the following error:

    Error: getaddrinfo ENOTFOUND domain domain:80

**Issue 2**
If Postman is setup to use System proxy and the proxy requires authentication, Postman will not prompt for credentials.

Following errors will be seen in DevTools console

    net::ERR_TUNNEL_CONNECTION_FAILED
    the server responded with a status of 407 (Proxy Authentication Required)

### **To Reproduce**
Steps to reproduce the behaviour:
**Issue 1**
1. Setup a global proxy 
2. Set username in the format `domain\username`
3. Try a request
4. In Postman Console / Dev-Console you can see that Postman tries to resolve the `domain` component and fails.

**Issue 2**
1. Setup Postman to use System Proxy
2. If this proxy requires auth, Postman would not request credentials
3. The request would thus fail.

### **Expected behaviour**
**Issue 1**: When the global proxy is set to use `domain\username` format, do not try to resolve the domain.
**Issue 2**: When System Proxy is set, there should be an option to set credentials which may automatically be used if the proxy requires auth, or a prompt to enter credentials should be provided.

### **Screenshots**
![image](https://user-images.githubusercontent.com/21366/69185144-0b7f5a80-0b7b-11ea-9f21-a517f211f89b.png)

<img width="1394" alt="Screen Shot 2019-11-20 at 9 51 47 AM" src="https://user-images.githubusercontent.com/21366/69185288-61540280-0b7b-11ea-8008-70be3aa64f43.png">

### **App information (please complete the following information):**
 - App Type : Native App
 - Postman Version : 7.9 Windows 2012 (I tested on OSX with 7.11)
 - OS: OSX, Windows 10, Winsows 2012 RC2

### **Additional context**
Logged by Postman for a report submitted internally.

**Comment by codenirvana** [comment]
`Issue 2` i.e, prompt to enter credentials is not supported right now.
The only way to configure proxy which requires authentication is via configuring Postman global proxy.

`Issue 1`, I am not sure about `domain\username` format 😕, Is it configured like this:

<img width="696" alt="Screenshot 2019-11-20 at 4 17 59 AM" src="https://user-images.githubusercontent.com/4865836/69193798-9cd7d800-0b4d-11ea-90a2-65bbddf4b91c.png">
?

**Comment by xk0der** [comment]
@codenirvana - Thanks for looking into this.

The Proxy Server "`hostname`" and the "`domain`" component in the username field may be different

`domain\username` is a format used primarily by MS Windows OS
Where the domain is a NetBIOS domain.
Reference: https://docs.microsoft.com/en-us/windows/win32/secauthn/user-name-formats

I had checked with the user if `UPN`  may work, which is in the format `username@domain`
But unfortunately, their proxy did not support that format.

*TL;DR*
This is how things will be configured
Proxy Server: `proxyHostname` : `ProxyPort`
Username: `domain\username`

**Comment by codenirvana** [comment]
@xk0der As a workaround, can you check if encoding the `\` to `%5C` in the `username` fixes this issue?

**Comment by xk0der** [comment]
@codenirvana : I did have the user try that and that did not work, unfortunately.

**Comment by saswatds** [comment]
@xk0der Is the proxy-server being used by any chance an NTLM proxy? We do not support NTLM proxy as of now. Tracking the same as a feature request here https://github.com/postmanlabs/postman-app-support/issues/3692

**Comment by Camios** [comment]
@codenirvana for `Issue 2`, is there a feature request to enable prompting or text boxes to provide [domain\]username and password for a system proxy? Does postman support pac files?
Also why aren't my HTTP_PROXY environment variables recognised when I use a batch file to set them and then start postman?

@saswatds I am double checking with the network team regarding proxy auth type.

**Comment by xk0der** [comment]
@saswatds Initial interactions internally had suggested that NTML was not in use, but I see that @Camios  has joined the discussion here and I'll let them confirm that.

**Comment by Camios** [comment]
@saswatds @xk0der, I have just heard back from the network team and found that they are using a proxy that supports multiple challenge types (kerberos, NTLM, basic) with a fallback mechanism from best to worst. If the application (e.g. Postman) doesn't support the higher it will drop to the weaker challenge type. I installed fiddler and it shows that some of the tunnel attempts include NTLM proxy authorization. So it appears the proxy is only falling back to NTLM challenge type and that feature request #3692 is the problem. How does postman team prioritise work on a feature that is fundamental to corporate environments, but the request is now more than 2yr-old?

**Comment by JorgeMerchan** [comment]
+1

**Comment by PeterMue** [comment]
I'm facing basically the same problem: our corporate proxy only supports kerberos or ntlm. Since this is a general problem for many applications we have a second proxy dedicated to our developers that is only accessible from theirs vlan and "authenticates" by ip. So thats our common workaround for ntlm proxy issues but, even this workaround is not that much fun with Postman:

from my understanding, postman differentiates between the "default proxy" and the "proxy for sending requests".
The proxy for sending requests is configurable and works as i would expect.

The default proxy however has no configuration options exception the authentication settings and uses by default (at least on macos) the system wide proxy settings (but not the environment variables HTTP_PROXY, HTTPS_PROXY!).
In my case, thats not practical, because my system wide settings use a corporate proxy with ntlm authentication (which is not supported by postman i think). Therefor i have a separate corporate proxy with ip based authentication configured using the enviroment variables.
However, since postman ist not respecting the environment variables for the "default proxy" stuff, which includes the login to postman cloud, it's not working.

The workaround for me is, to specify the proxy as argument when starting postman by doing something like 
`open -b com.postmanlabs.mac --args --proxy-server=http://my-corporate-proxy:port`.

**Describe the solution you'd like**
It would be great, if either the "default proxy" respects the environment variables and/or the settings would be extended to work the same manner as the "proxy for request" settings.




I agree with @Camios , the described proxy setup is very common in corporate environments but postman has no good solution yet.
When searching through the postman issues looking for proxy related stuff, there are more tickets that already mention this behaviour or suggest a solution.

**Comment by postman-kenji** [comment]
Closing this issue given workarounds provided and open feature request https://github.com/postmanlabs/postman-app-support/issues/3692.
